datafusion-examples/examples/relation_planner/pivot_unpivot.rs

```diff
@@ -0,0 +1,543 @@
+
+    // Register sample data tables
+    register_sample_data(&ctx)?;
+
+    // Register custom planner
+    ctx.register_relation_planner(Arc::new(PivotUnpivotPlanner))?;
+
+    println!("Custom Relation Planner: PIVOT and UNPIVOT Operations");
+    println!("======================================================\n");
+
+    // Example 1: Basic PIVOT to transform monthly sales data from rows to columns
+    // Shows: How to pivot sales data so each quarter becomes a column
+    // The PIVOT is rewritten to: SELECT region, SUM(CASE WHEN quarter = 'Q1' THEN amount END) as Q1,
+    //                             SUM(CASE WHEN quarter = 'Q2' THEN amount END) as Q2
+    //                             FROM quarterly_sales GROUP BY region
+    // Expected Output:
+    // +--------+------+------+
+    // | region | Q1   | Q2   |
+    // +--------+------+------+
+    // | North  | 1000 | 1500 |
+    // | South  | 1200 | 1300 |
+    // +--------+------+------+
+    run_example(
+        &ctx,
+        "Example 1: Basic PIVOT - Transform quarters from rows to columns",
+        r#"SELECT * FROM quarterly_sales
+           PIVOT (
+             SUM(amount)
+             FOR quarter IN ('Q1', 'Q2')
+           ) AS pivoted"#,
+    )
+    .await?;
+
+    // Example 2: PIVOT with multiple aggregate functions
+    // Shows: How to apply multiple aggregations (SUM and AVG) during pivot
+    // Expected: Logical plan showing MiniPivot with both SUM and AVG aggregates
+    // Actual (Logical Plan):
+    // Projection: pivoted.region, pivoted.Q1, pivoted.Q2
+    //   SubqueryAlias: pivoted
+    //     MiniPivot aggregate=[SUM(amount), AVG(amount)] value_column=[quarter] values=["Q1", "Q2"]
+    //       Values: (Utf8("North"), Utf8("Q1"), Int64(1000)), (Utf8("North"), Utf8("Q2"), Int64(1500)), ...
+    run_example(
+        &ctx,
+        "Example 2: PIVOT with multiple aggregates (SUM and AVG)",
+        r#"SELECT * FROM quarterly_sales
+           PIVOT (
+             SUM(amount), AVG(amount)
+             FOR quarter IN ('Q1', 'Q2')
+           ) AS pivoted"#,
+    )
+    .await?;
+
+    // Example 3: PIVOT with additional grouping columns
+    // Shows: How pivot works when there are multiple non-pivot columns
+    // The region and product both appear in GROUP BY
+    // Expected Output:
+    // +--------+-----------+------+------+
+    // | region | product   | Q1   | Q2   |
+    // +--------+-----------+------+------+
+    // | North  | ProductA  | 500  |      |
+    // | North  | ProductB  | 500  |      |
+    // | South  | ProductA  |      | 650  |
+    // +--------+-----------+------+------+
+    run_example(
+        &ctx,
+        "Example 3: PIVOT with multiple grouping columns",
+        r#"SELECT * FROM product_sales
+           PIVOT (
+             SUM(amount)
+             FOR quarter IN ('Q1', 'Q2')
+           ) AS pivoted"#,
+    )
+    .await?;
+
+    // Example 4: Basic UNPIVOT to transform columns back into rows
```

Comment:
Per your other comments I am not sure how important it is to show all these corner cases working, as I don't think they add much to the example, but they do help to give confidence to anyone who might copy/paste this code into their system